File: 4_go/ParentesisBalanceados.go
package main
import "fmt"
var apertura = map[rune]bool{
  '(': true,
  '[': true,
  '{': true,
}
var cierre = map[rune]bool{
  ')': true,
  ']': true,
  '}': true,
}
var comparador = map[rune]rune{
  ')': '(',
  ']': '[',
  '}': '{',
}

func ParentesisBalanceados(cadena string) bool {
  pila := []rune{}
  for _, caracter := range cadena {
    if _, encontrado := apertura[caracter]; encontrado {
      pila = append(pila, caracter)
      continue
    }
    if _, encontrado := cierre[caracter]; encontrado {
      if len(pila) == 0 {
	return false
      }
      if pila[len(pila)-1] == comparador[caracter] {
	pila = pila[0 : len(pila)-1]
      } else {
	return false
      }
    }
  }
  return len(pila) == 0
}

func main() {
  fmt.Println(ParentesisBalanceados("{[()]}"))
  fmt.Println(ParentesisBalanceados("{[(]}"))
}

File: 8_php/parentesis_balanceados.php
<?php
function parentesis_balanceados($cadena){
  $apertura = "([{";
  $cierre = ")]}";
  $comparador = array(
    ")" => "(",
    "]" => "[",
    "}" => "{"
  );
  $pila = [];
  $cadena = str_split($cadena);
  foreach($cadena as $caracter){
    if (str_contains($apertura,$caracter)){
      array_push($pila,$caracter);
    }
    elseif(str_contains($cierre,$caracter)){
      if (count($pila) == 0){
	return False;
      }
      if ($pila[count($pila) - 1] == $comparador[$caracter]){
	array_pop($pila);
      } else {
	return False;
      }
    }  
  }
  return count($pila) == 0;
}

echo parentesis_balanceados("{[()]}");
echo parentesis_balanceados("{[(]}");
?>
